File: examples/code/integrations/microsoft/custom_auth.py
from arcade.client.errors import APITimeoutError
from arcade.client import Arcade, AuthProvider

client = Arcade()

user_id = "user@example.com"

# Start the authorization process
auth_response = client.auth.authorize(
    provider=AuthProvider.microsoft,
    scopes=["User.Read", "Files.Read"],
    user_id=user_id,
)

if auth_response.status != "completed":
    print("Please complete the authorization challenge in your browser:")
    print(auth_response.auth_url)

while auth_response.status != "completed":
    try:
        auth_response = client.auth.status(auth_response, wait=60)
    except APITimeoutError:
        continue

token = auth_response.context.token
# Do something interesting with the token...

File: examples/code/integrations/microsoft/custom_auth.js
import { Arcade } from '@arcadeai/arcadejs';

const client = new Arcade();

const userId = '{arcade_user_id}';

// Start the authorization process
let authResponse = await client.auth.start(userId, 'microsoft', {
  scopes: ['User.Read', 'Files.Read'],
});

if (authResponse.status !== 'completed') {
  console.log('Please complete the authorization challenge in your browser:');
  console.log(authResponse.url);
}

// Wait for the authorization to complete
authResponse = await client.auth.waitForCompletion(authResponse);

const token = authResponse.context.token;
// TODO: Do something interesting with the token...
